[Spring 경로 조회 - 1단계] 승팡(김승래) 미션 제출합니다.

Seungpang opened the pull request:
안녕하세요 소니!
승팡입니다😀

테스트 코드를 사용할 때 저는 픽스처를 사용해서 각각 테스트별로 생성해서
테스트들이 각각 독립적이게 했는데 점점 테스트 때 생성되어야 할게 너무 많아지다 보니까
service-test-ddl.sql과 같은 파일로 데이터들을 넣어놓고 테스트하는 게 편해지더라고요
하지만 이럴 경우 ddl의 값이 변경될 경우 기존 테스트들이 다 깨질 수 있다는 문제점이 있습니다.
즉 저 ddl문에 종속적인 테스트라는 단점이 존재한다는 것은 인지합니다. 
그러나 이러한 테스트를 생성해야 할 로직이 점점 많아진다면 픽스처를 활용하는 게 최선일까요?

리뷰에 미리 감사드리고 제가 놓친 부분이나 고쳤으면 하는 부분이 있으면 적극 반영하겠습니다.
감사합니다!

Seungpang committed:
feat: 지하철 노선도 소스 동기화

Seungpang committed:
docs: 기능 요구 사항 작성

Seungpang committed:
feat: PathGraph 구현

Seungpang committed:
docs: 기능 요구 사항 추가

Seungpang committed:
feat: 최단 경로 조회 구현

Seungpang committed:
test: LineService 테스트 케이스 수정

Seungpang committed:
refactor: 코드 포맷 및 사용하지 않는 부분 제거

1. Path 매직넘버 상수화
2. 사용하지 않는 코드 제거
3. 포맷 정렬

Seungpang committed:
docs: 요구 사항 최신화

sonypark commented on src/main/java/wooteco/subway/service/PathService.java:
```diff
@@ -0,0 +1,43 @@
+
+        Station sourceStation = findStationById(source);
+        Station targetStation = findStationById(target);
+        Path path = pathGraph.findShortestPath(sourceStation, targetStation);
```
sections에 포함되지 않는 역의 path를 조회하면 어떻게 되나요?

Seungpang replied:
graph의 vertex에 포함되지 않아서 IllegalArgumentException이 발생합니다.
다만 그 부분이 외부 라이브러리 부분이라 이 예외를 도메인에 맞게 한번 감쌀 필요성이 있을 것 같습니다.
따라서 PathNotFoundException으로 해당 도메인에 맞게 메시지를 변경했습니다.

sonypark commented on src/test/java/wooteco/subway/domain/PathTest.java:
```diff
@@ -0,0 +1,22 @@
+
+@SuppressWarnings("NonAsciiCharacters")
+class PathTest {
+
+    @ParameterizedTest(name = "거리가 {0}일 때, 요금이 {1}원이다.")
+    @CsvSource(value = {"10,1250", "50,2050", "90,2550"})
+    void 요금을_계산한다(int distance, int fare) {
+        Station 강남역 = new Station("강남역");
```
경계값 테스트를 해주면 좋을 것 같아요
(10km, 11km, 50km, 51km)

Seungpang replied:
경계값에 대한 테스트를 해야지 생각하고 딱 경계부분만 테스트를 해버렸네요ㅠㅠ 추가했습니다.

sonypark commented on src/main/java/wooteco/subway/domain/PathGraph.java:
```diff
@@ -0,0 +1,44 @@
+import org.jgrapht.alg.shortestpath.DijkstraShortestPath;
+import org.jgrapht.graph.DefaultWeightedEdge;
+import org.jgrapht.graph.WeightedMultigraph;
+
+public class PathGraph {
+
+    private final WeightedMultigraph<Station, DefaultWeightedEdge> graph = new WeightedMultigraph<>(
+        DefaultWeightedEdge.class);
+
+    public PathGraph(List<Line> lines) {
+        initGraph(lines);
+    }
+
```
이 클래스는 테스트가 없네요🥲

Seungpang replied:
해당 클래스에 대한 테스트 케이스 추가했습니다.

sonypark commented on src/test/java/wooteco/subway/acceptance/PathAcceptanceTest.java:
```diff
@@ -0,0 +1,52 @@
+            post(makeSectionJson(2L, 3L, 3), "/lines/2/sections");
+
+            ExtractableResponse<Response> response = get(
+                "/paths?source=1&target=3&age=15");
+            PathResponse actual = response.jsonPath().getObject(".", PathResponse.class);
+
+            assertAll(() -> {
+                assertThat(response.statusCode()).isEqualTo(HttpStatus.OK.value());
+                assertThat(actual)
+                    .extracting("distance", "fare")
+                    .containsExactly(6, 1250);
+            });
```
응답값 검증 👍

sonypark commented on src/main/java/wooteco/subway/dto/PathResponse.java:
```diff
@@ -0,0 +1,37 @@
+        this.distance = distance;
+        this.fare = fare;
+    }
+
+    public static PathResponse of(Path path) {
+        List<StationResponse> stations = path.getStations().stream()
+            .map(StationResponse::of)
+            .collect(Collectors.toList());
+        return new PathResponse(stations, path.getDistance(), path.calculateFare());
+    }
```
정적 팩토리 메서드 사용 👍

sonypark reviewed (changes requested):
승팡 안녕하세요~
경로 조회 미션 잘 구현해주셨네요!👍
간단한 코멘트 남겼으니 확인 부탁드릴게요 :)

> 테스트 코드를 사용할 때 저는 픽스처를 사용해서 각각 테스트별로 생성해서
테스트들이 각각 독립적이게 했는데 점점 테스트 때 생성되어야 할게 너무 많아지다 보니까
service-test-ddl.sql과 같은 파일로 데이터들을 넣어놓고 테스트하는 게 편해지더라고요
하지만 이럴 경우 ddl의 값이 변경될 경우 기존 테스트들이 다 깨질 수 있다는 문제점이 있습니다.
즉 저 ddl문에 종속적인 테스트라는 단점이 존재한다는 것은 인지합니다.
그러나 이러한 테스트를 생성해야 할 로직이 점점 많아진다면 픽스처를 활용하는 게 최선일까요?

테스트 데이터는 가능한 해당 테스트를 위한 픽스처를 만들어서 사용하고, 셋팅해야할 테스트 데이터가 너무 많거나 여러곳에서 중복 사용하는 부분이 많아지면 승팡이 하신것처럼 sql 파일로 분리하는 방법도 많이 사용합니다. 다만 하나의 테스트 sql 파일이 여러개의 테스트에 영향을 미치게 되면 유지보수가 힘들고 데이터 변경시 영향범위도 커지기 때문에 sql 파일도 최대한 테스트 목적에 맞게 분리해서 작성하고(ex. service-test-ddl.sql에서 section 연결 sql 분리) 필요한 테스트 메서드 또는 클래스에서 `@Sql`로 가져와 사용하는게 좋은 것 같습니다.

Seungpang committed:
test: Path 요금계산 경계값 테스트 케이스 추가

Seungpang committed:
test: PathGraph테스트 추가

Seungpang commented:
안녕하세요 소니!
승팡입니다😀

최대한 빠르게 피드백을 반영해봤습니다.
외부 라이브러리를 사용할 때 발생할 수 있는 예외에 대해서 놓쳤던 것 같습니다 ㅠㅠ
발생할 수 있는 예외들을 도메인에 맞게 메시지를 재정의 하도록 수정했습니다.

리뷰에 미리 감사드리고 제가 놓친 부분이나 고쳤으면 하는 부분이 있으면 적극 반영하겠습니다.
감사합니다!

Seungpang committed:
refactor: 객체 생성 로직 dto로 이동 및 코드 정리

1. 객체 생성 책이 dto로 이동
2. StationDao 상수 부분 대문자로 변경
3. PathGraph graph 초기화 생성자로 이동
4. Section Integer사용하는 부분 int로 변경

Seungpang committed:
refactor: 최단경로 조회 예외 케이스 추가

sonypark reviewed (approved):
승팡 안녕하세요~
리뷰 빠르게 반영해주셨네요💯
이번 단계는 여기서 머지해도 될 것 같아요.
다음 단계 진행 부탁드려요~~
